client/src/main/java/io/confluent/kafka/schemaregistry/AbstractSchemaProvider.java

```diff
@@ -61,27 +61,27 @@ private void resolveReferences(Schema schema, Map<String, String> schemas, Set<S
           || reference.getVersion() == null) {
         throw new IllegalStateException("Invalid reference: " + reference);
       }
+      QualifiedSubject refSubject = QualifiedSubject.qualifySubjectWithParent(
+              schemaVersionFetcher().tenant(), schema.getSubject(), reference.getSubject());
+      Schema s = schemaVersionFetcher().getByVersion(refSubject.toQualifiedSubject(),
+              reference.getVersion(), true);
+      if (s == null) {
+        throw new IllegalStateException("No schema reference found for subject \""
+                + refSubject
+                + "\" and version "
+                + reference.getVersion());
+      }
+      if (reference.getVersion() == -1) {
+        // Update the version with the latest
+        reference.setVersion(s.getVersion());
+      }
       if (visited.contains(reference.getName())) {
         continue;
       } else {
         visited.add(reference.getName());
       }
+      resolveReferences(s, schemas, visited);
```

Comment:
Can this line be moved to right after the `if` statement (where it was before)?

Reply:
No. That would cause circular reference to get into infinite recursion.

Reply:
I was suggesting moving it down, so that it looks like
 ``` if (!schemas.containsKey(reference.getName())) {
        resolveReferences(s, schemas, visited);
        schemas.put(reference.getName(), s.getSchema());
      }```

Reply:
Oh. I thought you meant after if(visited). That should be ok as we should have got these reference and updated the version. I will double check with the test cases.